session/pingpong/consumer_balance_tracker.go

```diff
@@ -211,43 +190,48 @@ func (cbt *ConsumerBalanceTracker) decreaseBalance(id identity.Identity, b uint6
 	}
 }
 
-func (cbt *ConsumerBalanceTracker) updateBCBalance(id identity.Identity, bcBalance, accountantPromised uint64) {
+func (cbt *ConsumerBalanceTracker) updateBalanceFromAccountant(id identity.Identity) {
+	cd, err := cbt.accountantBalanceFetcher(id.Address)
+	if err != nil {
+		log.Error().Err(err).Msg("could not get accountant balance")
+		return
+	}
+
 	cbt.balancesLock.Lock()
 	defer cbt.balancesLock.Unlock()
 	if v, ok := cbt.balances[id]; ok {
-		before := v.CurrentEstimate
+		isIncreased := true
 		var diff uint64
-		if bcBalance >= v.BCBalance {
-			if bcBalance-v.BCBalance >= accountantPromised {
-				diff = bcBalance - v.BCBalance - accountantPromised
-			} else {
-				log.Error().Msgf("uint64 underflow for balance calculations. BCBalance %v, currentBalance %v, accountantPromised %v", bcBalance, v.BCBalance, accountantPromised)
+
+		if v.BCBalance >= cd.Balance {
+			isIncreased = false
+			diff = safeSub(v.BCBalance, cd.Balance)
+		} else {
+			diff = safeSub(cd.Balance, v.BCBalance)
+		}
+
+		if isIncreased {
+			cbt.balances[id] = Balance{
+				BCBalance:       v.BCBalance + diff,
+				CurrentEstimate: v.CurrentEstimate + diff,
 			}
-			v.BCBalance += diff
-			v.CurrentEstimate += diff
 		} else {
-			v.BCBalance = 0
-			v.CurrentEstimate = 0
+			cbt.balances[id] = Balance{
+				BCBalance:       safeSub(v.BCBalance, diff),
+				CurrentEstimate: safeSub(v.CurrentEstimate, diff),
+			}
 		}
-		cbt.balances[id] = v
-		go cbt.publishChangeEvent(id, before, v.CurrentEstimate)
 	} else {
 		cbt.balances[id] = Balance{
-			BCBalance:       bcBalance - accountantPromised,
-			CurrentEstimate: bcBalance - accountantPromised,
+			BCBalance:       cd.Balance,
+			CurrentEstimate: cd.Balance,
 		}
-		go cbt.publishChangeEvent(id, 0, bcBalance-accountantPromised)
 	}
 }
 
-func (cbt *ConsumerBalanceTracker) getBCBalance(id identity.Identity) (uint64, error) {
-	addr, err := cbt.channelAddressCalculator.GetChannelAddress(id)
-	if err != nil {
-		return 0, err
+func safeSub(a, b uint64) uint64 {
```

Comment:
Is is safe to ignore possible underflows with money? What if there is a bug and balance was decreased by more than it supposed to?

Reply:
The truth is always in the blockchain and there no underflows or overflows will happen. This is just a counter that doesn't do much really, just show balance estimates.